File: 3058-maximum-number-of-k-divisible-components/maximum-number-of-k-divisible-components.cpp
class Solution {
    int res=0;
    int dfs(vector<vector<int>> &adj,vector<int> &values,int k,int curr,int prev){
        long long sum=0;
        for(int i:adj[curr]){
            if(i==prev) continue;
            int x=dfs(adj,values,k,i,curr);
            if(x%k==0) res++;
            else sum+=x;
        }
        return (sum+values[curr])%k;
    }
public:
    int maxKDivisibleComponents(int n, vector<vector<int>>& edges, vector<int>& values, int k) {
        vector<vector<int>> adj(n);
        for(auto e:edges){
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
        }
        int x=dfs(adj,values,k,0,0);
        if(x%k==0) res++;
        return res;
    }
};

File: 3058-maximum-number-of-k-divisible-components/maximum-number-of-k-divisible-components.py
from typing import List

class Solution:
    def __init__(self):
        self.res = 0
        self.adj = []
        self.values = []
        self.k = 0
    @cache
    def dfs(self, curr: int, prev: int) -> int:
        total = 0
        for neighbor in self.adj[curr]:
            if neighbor == prev:
                continue
            x = self.dfs(neighbor, curr)
            if x % self.k == 0:
                self.res += 1
            else:
                total += x
        return total + self.values[curr]

    def maxKDivisibleComponents(self, n: int, edges: List[List[int]], v: List[int], ki: int) -> int:
        self.adj = [[] for _ in range(n)]
        self.values = v
        self.k = ki

        for u, v in edges:
            self.adj[u].append(v)
            self.adj[v].append(u)

        ans = 0
        for i in range(n):
            self.res = 0
            x = self.dfs(i, i)
            if x % self.k == 0:
                self.res += 1
            ans = max(ans, self.res)
        return ans
